plugins/query-insights/src/main/java/org/opensearch/plugin/insights/rules/model/SearchQueryRecord.java

```diff
@@ -0,0 +1,215 @@
+        this.searchType = SearchType.fromString(in.readString().toLowerCase(Locale.ROOT));
+        this.source = in.readString();
+        this.totalShards = in.readInt();
+        this.indices = in.readStringArray();
+        this.propertyMap = in.readMap();
+    }
+
+    public SearchQueryRecord(
+        final Long timestamp,
+        final SearchType searchType,
+        final String source,
+        final int totalShards,
+        final String[] indices,
+        final Map<String, Object> propertyMap,
+        final T value
+    ) {
+        this(timestamp, searchType, source, totalShards, indices, propertyMap);
+        this.value = value;
+    }
+
+    public SearchQueryRecord(
+        final Long timestamp,
+        final SearchType searchType,
+        final String source,
+        final int totalShards,
+        final String[] indices,
+        final Map<String, Object> propertyMap
+    ) {
+        this.timestamp = timestamp;
+        this.searchType = searchType;
+        this.source = source;
+        this.totalShards = totalShards;
+        this.indices = indices;
+        this.propertyMap = propertyMap;
+    }
+
+    /**
+     * The timestamp of the top query.
+     */
+    public Long getTimestamp() {
+        return timestamp;
+    }
+
+    /**
+     * The manner at which the search operation is executed.
+     */
+    public SearchType getSearchType() {
+        return searchType;
+    }
+
+    /**
+     * The search source that was executed by the query.
+     */
+    public String getSource() {
+        return source;
+    }
+
+    /**
+     * Total number of shards as part of the search query across all indices
+     */
+    public int getTotalShards() {
+        return totalShards;
+    }
+
+    /**
+     * The indices involved in the search query
+     */
+    public String[] getIndices() {
+        return indices;
+    }
+
+    /**
+     * Get the value of the query metric record
+     */
+    public T getValue() {
+        return value;
+    }
+
+    /**
+     * Set the value of the query metric record
+     */
+    public void setValue(T value) {
+        this.value = value;
+    }
+
+    /**
+     * Extra attributes and information about a search query
+     */
+    public Map<String, Object> getPropertyMap() {
+        return propertyMap;
+    }
+
+    @Override
+    public int compareTo(SearchQueryRecord<T> otherRecord) {
+        return value.compareTo(otherRecord.getValue());
+    }
+
+    public boolean equals(SearchQueryRecord<T> other) {
+        if (false == this.timestamp.equals(other.getTimestamp())
+            && this.searchType.equals(other.getSearchType())
+            && this.source.equals(other.getSource())
+            && this.totalShards == other.getTotalShards()
+            && this.indices.length == other.getIndices().length
+            && this.propertyMap.size() == other.getPropertyMap().size()
+            && this.value.equals(other.getValue())) {
+            return false;
+        }
```

Comment:
This condition looks wrong. 

It's saying "If the timestamps are not equal, but all of these other fields **are** equal, then the two records are not equal.

Should it be:

```
        if (false == this.timestamp.equals(other.getTimestamp())
            || !this.searchType.equals(other.getSearchType())
            || !this.source.equals(other.getSource())
            || this.totalShards != other.getTotalShards()
            || this.indices.length != other.getIndices().length
            || this.propertyMap.size() != other.getPropertyMap().size()
            || !this.value.equals(other.getValue())) {
            return false;
        }
```

Reply:
Good catch.. Just resolved this dumb mistake.